### ISSUE TITLE
Large file support

### ISSUE THREAD

**Issue by watsonian** [atom/atom#307]
So, I use my text editor for all kinds of things (writing code, reading readmes, editing dotfiles, etc). One of the things I do most often during any given day is open up large log files to troubleshoot Enterprise problems. These files can sometimes get as large as 500-800MB.

I just tried viewing a 350MB log file and Atom locked up immediately -- the file selector didn't change and the entire window went completely white 4 seconds after I tried viewing the file. I could still close the top level window, but we should still have better handling for this kind of thing.

Sublime Text 2 took about 55 seconds to open the file, but gave a nice progress indicator while it was loading:

![Screen Shot 2013-02-24 at 4 50 56 PM](https://f.cloud.github.com/assets/244/190379/c0abc27e-7ee5-11e2-8ec4-4205cdee0ad0.png)

Once it loaded it was as responsive as any other file I load (scrolled fast, normal text highlight performance, etc). It would be nice if we set this as a baseline for expected behavior (progress bar + responsive after loading).

**Comment by LevelTranslation** [comment]
I want to use Atom to edit books in markdown. Large files might be an edge case for some, but for me, it's my bread and butter. I want to get away from vim, because looking at prose in monospaced fonts is painful, but it continues to be the only thing that will reliably open a whole book.

**Comment by kenokabe** [comment]
>Can't even open files larger than 2 MB. 

Confirmed here, too. Cannot open some [webpack](https://github.com/webpack/webpack)ed large file.

>This is a major deal for us preventing switch to Atom.
I agree. Even I already have switched to Atom for a while now, it's a matter of major quality as an editor.

**Comment by mrmylanman** [comment]
I just want to throw in my $0.02, as well. This is a feature that I would really like. At the company I work for we import some very large text files regularly and when something goes wrong I often have to open the file to verify the format hasn't changed, data isn't corrupted, and that particular lines are what I expect.

Sometimes I will have to edit these files by hand, so write support would be beneficial.

I don't expect it to be fast with such a large file (Sublime certainly isn't), but having it at least work would be an improvement over the present situation.

These files are routinely as large as 10-20 MB in size. This is literally the only reason why I have Sublime Text installed still, as for everything else I prefer Atom.

Would be great to see large file support!

**Comment by AbdealiJK** [comment]
I too need this feature, i frequently open sql files and log files

**Comment by telkomsel666** [comment]
#1
12 Jan 2015 22:49, "Mylan Connolly" <notifications@github.com> menulis:

**Comment by mark-hahn** [comment]
>  i frequently open sql files and log files

If you only need to read them try out my view-tail-large-files package.

**Comment by ilanbiala** [comment]
@mark-hahn I'm sure many people greatly appreciate the package, but this is really something that has to be part of the core editor. The Atom team can try to integrate and use parts of your package, which may be a good idea, but they shouldn't ignore this issue just because a package exists. Right now, most people prefer Sublime because it is quicker, more stable, and better designed. Atom will likely improve in the latter two as releases are made and more feedback is given, but the first will be tricky due to the program being written in CS/JS.

**Comment by blakegarretson** [comment]
Not to beat a dead horse, but I'll add another datapoint and lend some support to the "large file" camp. I edit (not just view) files in the 10MB-200MB range every day.  This is common in the scientific computing community, specifically fields like FEA and CFD where input files are large and often require hand editing.  I would love to use Atom on FEA input files if it could handle them.  I wrote a package to support Abaqus (FEA code) syntax in Atom, and then was promptly disappointed to find almost none of my files would open due to the size limitation.

**Comment by mark-hahn** [comment]
> they shouldn't ignore this issue just because a package exists.

Of course not.  I was just offering an option you can use now.  Also it
also offers many features that shouldn't be in core (actually too many
features).

P.S.  From my experience core has never been shy to add new features
because of existing packages.

**Comment by demisx** [comment]
Atom 1) being slow and 2) not being able to support large files are the 2 main reasons why we can't switch from SublimeText yet.

**Comment by simonzack** [comment]
> Atom 1) being slow and 2) not being able to support large files are the 2 main reasons why we can't switch from SublimeText yet.

This is precisely the same reason I need to use sublime text instead of atom.

**Comment by mark-hahn** [comment]
Atom is always going to be slower than ST because it uses Node instead of
C++.  The advantage of using node is that packages can be easily added by
anyone who knows web development.  The feature set of Atom will grow
rapidly and leave the other programming editors in the dust, including ST.

I personally use notepad++ for the occasions I need to quickly open a
simple file.  I keep an Atom window open all the time for each project I'm
developing.  Editing a file and program development are quite different
endeavors.

If you insist on using one editor for all editing then Atom will probably
never be right for you.

**Comment by simonzack** [comment]
> Atom is always going to be slower than ST because it uses Node instead of C++.

I agree, but it doesn't have to be so slow to the degree that you're constantly aware of it. I think this is achievable. Look at brackets, for example.

**Comment by demisx** [comment]
Agreed. It can be slower, but not slow. Also, the 2MB file limit is a must fix for any text editor.

**Comment by mark-hahn** [comment]
> Also, the 2MB file limit is a must fix for any text editor.

This is something that can be fixed unlike the speed problem.  However it will be a really tough nut to crack.  It would require a major refactoring of a large part of Atom.  It certainly won't be part of 1.0 but might be in 2.0 or 3.0. 

>  is a must fix for any text editor.

But not for a programming editor. If your source files are over 2MB then you really need to rethink your app's architecture.

**Comment by demisx** [comment]
@mark-hahn There are concatenated files that exceed 2MB. Has nothing to do with the app's arch.

**Comment by benogle** [comment]
Just to be clear, improving performance and supporting larger files are both things within reach and both part of our 1.0 focus.

**Comment by ilanbiala** [comment]
@benogle and the rest of the Atom team, have you guys looked at all into V8 bindings for the editor part to speed it up?

**Comment by benogle** [comment]
@ilanbiala I'm not following. Which bindings, do you have a link?

**Comment by ilanbiala** [comment]
@benogle http://nodejs.org/api/addons.html. I believe that you can still expose the necessary methods for packages and themes, but it should be faster than writing in CS/JS.

**Comment by benogle** [comment]
So you're asking if we've considered dropping down to c++ for some of it. We may consider it for some things, though it's not clear what makes sense to write in c++ at this time.

**Comment by toadwarrior** [comment]
I'd agree with Mark as to why Atom is slow but I'm not sure I can agree that any web developer being able to make add-ons is worth such a feature regression.

Many web developers, such as myself, can write in other languages. I can understand why someone may not want C but if you're working in JS enough to consider writing your own plugins then something like Java or Python shouldn't be hard to grasp. Both are "web" languages anyway.

of course there is already a definitive IDE written in Java (the JetBrains' offerings) so if someone were to choose to write yet another Java IDE / text editor they'd have to be good from the get go rather than relying on just having something new and shiny to mask the fact it's not very good when compared to the competition.

**Comment by mark-hahn** [comment]
>  I'm not sure I can agree that any web developer being able to make
add-ons is worth such a feature regression.

The node decision was made a long time ago.  It is the heart of Atom.
Having worked on Node and the browser for years this is an exciting
opportunity for me and many others.

>  than relying on just having something new and shiny to mask the fact
it's not very good when compared to the competition.

No one is masking anything.  Atom is advertised as a hackable editor and
it fulfills that promise as no other editor can.

>  it's not very good when compared to the competition.

It might not match your needs but it definitely meets many other's needs.
This is not a contest.  It isn't a competition. If it is the right tool
then use it.

**Comment by ilanbiala** [comment]
@benogle the main reason to drop down to C++ is to speed it up, so anything that is significantly slower than the rest of the program and bottleneck's user workflow is something that should be analyzed and possibly written in C++. The biggest thing everybody has an issue with right now is the editor view and that almost guaranteed needs to be in C++ to actually be fast and responsive with large files. Considering the fact that the API is already there, putting it into C++ shouldn't be so complicated.

**Comment by benogle** [comment]
Right, when we really start digging into it, the slow and memory intensive bits will make themselves evident, and we will decide what to do on a case by case basis.

**Comment by mark-hahn** [comment]
I originally compared C++ to Node, but the real slowdown in the actual
editor is the DOM.  The DOM is as integral to Atom as Node is.  I don't see
how writing additional C++ code can speed up DOM rendering.  Just as in web
browsers the DOM has a zillion positive features but speed is not one of
them.

I apologize for being negative.  I hope there is some benefit to my
comments.  Maybe setting expectations is good and productive?  Also I hope
I'm wrong but I doubt it in this case.

**Comment by ilanbiala** [comment]
@mark-hahn yes the DOM is also going to be very slow. We should actually see which takes more time, and optimize that first and see if it's enough until the rest of the editor catches up.

**Comment by demisx** [comment]
Not sure if this is DOM related, but Atom start-up time and switching between projects is very slow too.

**Comment by karai17** [comment]
To throw some fuel on the fire, I also often need to open "large" files. Just a few minutes ago I needed to open a 5MB IQE 3D model file and was a bit shocked that Atom told me it couldn't.

After reading through the history of this issue, I feel it is worth noting that Notepad++ supports large files. I learned this several years ago while I was bored in class and decided to write a C++ program that would count to 1 million and append each number to a new line in a file. When I tried opening up the .txt file with Windows Notepad, as you can imagine, things went poorly. Opening the same file in np++ had virtually no lag and I was able to interact with the file normally.

The reason I bring up np++ and not yammer on about Sublime is because np++ is open source (GPL). If you want to dig through their code to see how they open large files and perhaps use a similar method in Atom, I think that would be at least useful for research.

http://notepad-plus-plus.org/download/v6.7.4.html

It does seem odd to me that the general attitude towards large files by the Atom developers is to consider it an "edge case" and even attempt to ignore the issue. Large files are not an edge case. Logs, SQL dumps, markdown/epub/etc books, generated files, various exports... There are tonnes of reasons why someone would want to open a file that doesn't sit neatly within a few hundred KB. SQL is a great example for this since Atom is built on web standards and then immediately ignores a significant feature of the web itself: databases.

If Atom wants to succeed as the go-to text editor, it's going to need to take a step back and rethink the basics. If Atom can't open a text file for absolutely any reason, then Atom isn't ready to call itself a text editor. That may sound harsh, but I believe it to be true. Editing a text file is **the** fundamental purpose of a text editor, and we need to get that right before we can consider anything else.

All that being said, I am quite happy with all the other features of Atom. After poking it for a few hours I am already feeling ready to shift away from Sublime, but this single issue stopped me dead in my tracks. Let's hit this one out of the ballpark and solidify Atom as **the** text editor.

**Comment by Sacristan** [comment]
Im often using N++ to view COLLADA 3d format (*.dae) XML data which is really large. 
3d Mesh export data needs to be validated also 'by hand' sometimes.

With Atom it isnt possible. 

This surely is no Edge-case as a text editor WOULD need to support to open every kind of file.

**Comment by karai17** [comment]
I'd also like to point out that Sublime is also capable of opening binary files. While i don't necessarily suggest that Atom should focus their attention on opening arbitrary files, I don't see why it shouldn't be able to. Sublime's method even goes so far as to display binary files in hex, which is really cool.

I can already picture an Atom package that translates hex into ASCII. I used ot have a hex editor that showed both the hex and the ASCII, and it was pretty useful in some cases. Again, not exactly important, but I think opening arbitrary files and displaying their contents in a safe way would be neat.

**Comment by nathansobo** [comment]
> It does seem odd to me that the general attitude towards large files by the Atom developers is to consider it an "edge case" and even attempt to ignore the issue.

Atom developer here. I do consider it an edge case, but I think it's an important one and I don't intend to ignore it. Many concerns need consideration in deciding where to spend resources, but text editing in general is obviously high on the list and that includes supporting larger files.

> If you want to dig through their code to see how they open large files and perhaps use a similar method in Atom, I think that would be at least useful for research.

Sounds good. If you want this to happen faster, perhaps you could also give that a try and post a report here or somewhere else. This is probably too big an issue to solve in a single PR, but experimentation and information could be helpful.

**Comment by dalekprime** [comment]
I see JS and DOM mentioned as the challenges for improved performance.
I'm just wondering if anyone knows how lighttable or brackets have implemented their implementation.  In the case of lighttable, it can handle large files and seems just as responsive as sublime.

**Comment by martinlindhe** [comment]
I think binary viewing/editing should be outside the scope of atom. There are many tools for that. With that said, i would also like to point out the hex plugin for atom, that allows you to view files in binary: https://atom.io/packages/hex

Regarding huge file editing, I have had to deal with some huge files (200 mb + sql dumps or log files) and even sublime is having a hard time with these (Sublime seems to work fine for files in like 20-40 mb size). For this less common use case, i have resorted to use Textwrangler; while not open source it is freeware and does handle these huge files much better than Sublime appears to do.
Just my two cents.

**Comment by karai17** [comment]
@nathansobo While I have no problem helping with research, I must admit that my knowledge with C++ is extremely limited. I will take a peak at the source though and see if I can't find something useful.

**Comment by mark-hahn** [comment]
> Regarding huge file editing, I have had to deal with some huge files (200
mb + sql dumps or log files)

I probably mention this too often but if you only need read-only access try
my view-tail-large-files package.

**Comment by nathansobo** [comment]
> I probably mention this too often but if you only need read-only access try my view-tail-large-files package.

@mark-hahn I think it's worth repeating.

**Comment by karai17** [comment]
@nathansobo After taking a peak at the source, I think I found the file loading bit.

http://hastebin.com/irivodugos.cpp

From the looks of it, Notepad++ checks the file size and then loads the file in chunks (in this case 256 bytes) instead of trying to load the whole file at once.

Lua (which has very nice file loading) seems to do the same:

http://www.lua.org/source/5.2/liolib.c.html#read_all

I think the solution here is Atom needs to read files in chunks and assemble them on the fly in memory instead of trying to load up whole files at once.

**Comment by mark-hahn** [comment]
> assemble them on the fly in memory

It would have to only have a cache in memory since the whole file wouldn't
fit.  Then on every text load it would have to have all the css and grammar
stuff applied again.  This is non-trivial.  I know since I did it for my
view-tail-large-files package.

**Comment by LPGhatguy** [comment]
Is mmap/whatever the Windows equivalent a reasonable enough solution?

In other news, Sublime Text has a clever way of doing things when it comes to large files.

I used the following Lua script to generate 200MB files. The first part creates 200MiB of solid number signs ("#"), while the second creates 200MiB consisting of repeats of 1023 number signs followed by a newline:

```lua
do -- 200 MiB solid
	local handle = io.open("garbage.txt", "wb")
	local garbage = ("#"):rep(1024)
	local mb = 200

	for i = 1, 1024 * mb do
		handle:write(garbage)
	end
	handle:close()
end

do -- 200 MiB, newlines every kilobyte
	local handle = io.open("garbage-endl.txt", "wb")
	local garbage = ("#"):rep(1023) .. "\n"
	local mb = 200

	for i = 1, 1024 * mb do
		handle:write(garbage)
	end
	handle:close()
end
```

Sublime Text will not open the file without newlines, presumably because it buffers only up to a certain number of newlines, but it does successfully open the one with newlines placed every kilobyte.

Does that mean anything for an implementation?

**Comment by karai17** [comment]
I am not trying to suggest that this is a trivial issue, I am only
suggesting that the seemingly correct approach to loading files is to load
them in chunks. This will stop a 200mb file from taking up 5gb of ram and
never loading.

Load a chunk, apply metadata (styles, etc), load next chunk... If I am
going to try and open an 800mb log file, I should be expecting that will
take up at least 800mb of memory.


---
Landon Manning
lmanning17@gmail.com

**Comment by mark-hahn** [comment]
>  If I am going to try and open an 800mb log file, I should be expecting
that will take up at least 800mb of memory.

Have you ever looked at Atom in the inspector?  The dom probably takes ten
times the memory of the text alone.

**Comment by karai17** [comment]
No, honestly I've only been looking at Atom since last night. This
particular issue popped up very quickly in my testing since I very often
need to open files that are several megabytes in size.


---
Landon Manning
lmanning17@gmail.com

**Comment by toadwarrior** [comment]
If atom doesn't read in chunks as it were mentioned other NP++, I think that would be a sensible approach. I would have assumed that's what most editor do or at least they load the whole file into memory and then only apply the visual eye candy to certain chunks.

If there were a slight delay between move out of a current styled chunk of a file into the new chunk that would be an acceptable compromise and better than no option.

Many people will edit files that are easily under the 2mb 99% of the time. But I think everyone has had to edit larger files and if you have an editor you genuinely prefer and find it easier to work with it can be annoying to have to keep something else on hand for those other cases. I think larger files will more common in some technical fields. Some medical APIs I've dealt with return large chunks of data and I'll want to take copies of responses for testing.

Someone mentioned lighttable. That might be a good one to look at. I would assume its the most similar and some performance results rate it well for opening files ~20mb.

**Comment by mark-hahn** [comment]
If you make it truly disk-based then you can load **very** large files.  My package test uses a gigabyte file with ten million lines and it scrolls as fast as you can see.

**Comment by nathansobo** [comment]
The trick is implementing code folding, soft wrapping, syntax highlighting, etc for very large files.

**Comment by mark-hahn** [comment]
And it's going to be especially slow when you open a file that is
positioned at the bottom.

**Comment by waseemmachloy** [comment]
dont open file larger then 2mb at all.

**Comment by Sacristan** [comment]
Problem solving rule: Do not avoid the problem! It will just increase exponentially!

**Comment by 50Wliu** [comment]
It's not being avoided.  To quote @nathansobo from above:
> I think it's an important one and I don't intend to ignore it.

It's just that this is a large problem to tackle and not the only thing needed to get to Atom 1.0.

**Comment by mark-hahn** [comment]
You can open any sized file, read-only, with the view-tail-large-files
package.

**Comment by mnagel** [comment]
I had Atom fail to open a 2.1 MB file and was wondering if it wasn't better to show a message reading:

> Unfortunately Atom is very bad at processing large files. We are working on it, see issue #307! However, for the moment, we advise against opening any file larger than 2MB. It will probably crash Atom. It might even crash you computer!

And then have two buttons "Back to safety (do not open the file)" and "Proceed at my own risk". The former being the default.

That way users had the power to decide for themselves if they want to order their computer to try to handle the file.

**Comment by dalekprime** [comment]
I would prefer for now to make to have something like mark-hahn's package to be part of the editor.
1) Attempting to open a file larger than 2mb would then prompt you with something like
"This file is larger than 2mb, will be opened in read only mode with limited functionality..."
2) Allow users to change the default 2mb threshold.

As for future capabilities of large file support.  I think it is a different use case than normal editing.
Priority should first be on simply viewing and navigating the file.  Searching etc and highlighting are useful for log viewing.
Later, maybe the method for handling large files could be inline editing like is done is brackets.  So the file is opened read only.  You can only edit a section at at time where the editor will open that part as editable.

**Comment by kevinchiu** [comment]
Having large file support would be very handy when editing genetic data... about 1GB size for typical files.

**Comment by mark-hahn** [comment]
>  large file support would be very handy when editing genetic data

For specific things like that you should develop a package to show the data
in a cool formatted way with intelligent editing.  This is the kind of
stuff that makes Atom shine.

**Comment by batjko** [comment]
Ooh, a atom-genetic syntax pack! :microscope:

**Comment by Zireael07** [comment]
Some kind of an error message on trying to open >2 MB files would be helpful while you're working on fixing the underlying issue.

**Comment by lee-dohm** [comment]
@Zireael07 I'm pretty sure @benogle added some kind of error message in recent versions of Atom. He can correct me if I'm wrong though.

**Comment by benogle** [comment]
> I'm pretty sure @benogle added some kind of error message in recent versions of Atom. He can correct me if I'm wrong though.

Yeah there is a nicer error that points back to this issue since like v0.178 or so.

**Comment by backspaces** [comment]
Just a FYI: The GeoJson preview package is rendered pretty useless by this limit: most map files are quite large.

@mark-hahn makes a good point .. packages dealing with huge files, especially read-only files, might consider a strategy that converts to a nifty view.

The geojson package, however, was designed as a previewer of an existing file with its own editor for json files which are edited within atom.

**Comment by evandentremont** [comment]
+1.

Pasting ~2000 lines grinds to a halt on a modern machine. Using notepad or Mousepad or Textedit works faster than I can blink.

**Comment by WouterDS** [comment]
+1

**Comment by smiffy6969** [comment]
So I have mentioned this in a separate issue which has been closed (and pointed to this one).... I originally thought it was a linter (php) which I have disabled now, this has improved things a lot but I still get the issue, just takes longer to surface. The hangups I am getting, where the UI locks and or takes a good second per key stroke to update the UI seem to still be happening with pretty much all invasive packages disabled.

I have just noticed today that this seems to be happening after a few hours of use, and it has started happening, closing down and opening still seem crappy, but if I close machine and restart, things are good. This uses chrome right? that has a system process so any memory issue could still be present after restarting atom?

I love this editor, but I must have the ability to alter large files (even if I do not want to), I am resulting to going back to sublime to alter large files which is grating on me somewhat now.

Love the editor, hate this bug :(

working on new machine, plenty of RAM, SSD.......

**Comment by postcasio** [comment]
> This uses chrome right? that has a system process so any memory leak issue could still be present after restarting atom?

Quitting Atom will close all Atom processes, nothing should hang around.

Does the slow down also happen in [safe mode](https://atom.io/docs/v0.186.0/debugging#check-if-the-problem-shows-up-in-safe-mode)? Does it happen on only large files, or all files?

**Comment by mark-hahn** [comment]
> Quitting Atom will close all Atom processes

That has not been my experience but I haven't checked recently.

**Comment by sandstrom** [comment]
A wish would be 'inline-autocomplete', i.e. an UI-less mode where a keypress will rotate among the alternatives (similar to how it works in Textmate).

This package supports this, but it's not as nicely integrated with Atom as the regular autocomplete package, and since this is a core feature of an editor I'd much prefer using an official package.
https://atom.io/packages/inline-autocomplete

**Comment by octref** [comment]
Surprised this issue is not mentioned in GSOC's "Improve Atom's parser performance".

As @kevinsawicki suggested in [here](https://github.com/github/gsoc/issues/26)
> Here is where Atom starts chunking up lines to tokenized: 
https://github.com/atom/atom/blob/48d4ea0907e8aba68a8734484ceabdf3f0d02530/src/tokenized-buffer.coffee#L139-L144

However, in 99% scenario large files end up being log files, and there is no point tokenizing them.
Now Atom uses a `NullGrammar` to tokenize it. Basically going through the large file doing nothing.

So I added a few lines here:
```
    TokenizedBuffer.prototype.tokenizeInBackground = function() {
      if (!this.visible || this.pendingChunk || !this.isAlive()) {
        return;
      }

      // Not tokenizing on files Atom doesn't have grammar for
      if (this.grammar === this.grammar.registry.nullGrammar) {
        return;
      }

      this.pendingChunk = true;
      return _.defer((function(_this) {
        return function() {
          _this.pendingChunk = false;
          if (_this.isAlive() && _this.buffer.isAlive()) {
            return _this.tokenizeNextChunk();
          }
        };
      })(this));
    };
```

Take 2s to load a 580k log file. Scrolling and cursor-movement are smooth, but editing can be slow.
Atom used to take 5s to load this file on my machine, and another 20s before I can do basic scrolling with no lag, let alone editing.

Should I prepare a PR for this?
First time contributing to Atom here, so I hope @kevinsawicki or someone here can give me some feedback.

Meanwhile, I'll keep investigating why editing is slow. My guess is it's hooked on some expensive events.

---

@nathansobo 
> Yeah, we currently do a blocking read to load files. We need to switch to
something event-driven. I also wonder if we can load the file progressively
off the disk as needed so we don't have to seek the whole thing into
memory. 

Just wondering, how do you think of this idea as of now? Is it still worth doing or it'll break too many things? If it's worth it can I do it as a GSOC summer project?
And I'm curious as to your current plan/direction/progress for this issue. Do you need a hand?

**Comment by mark-hahn** [comment]
>  in 99% scenario large files end up being log files

I apologize if I've mentioned this recently, but the package
`view-tail-large-files` is perfect for logs.  Scrolling and tailing
gigabyte files is no problem.

**Comment by octref** [comment]
> I apologize if I've mentioned this recently, but the package
`view-tail-large-files` is perfect for logs.  Scrolling and tailing
gigabyte files is no problem.

I'm aware of that packge. However, it is read-only. And I don't think there is a conflict between "using that package temporarily for viewing large files" and "making Atom support large file natively".
Still, there is no point for Atom to tokenize log file, which takes around 20s for a 500kb-ish file while freezing the UI. This fix at least let Atom open & view large file natively without much hang.

**Comment by MatthewSteen** [comment]
Glad to see this is `in-progress`, I commonly need to open text files of building energy models in the 10 MB range.

**Comment by jeffschwartz** [comment]
+1

**Comment by nathansobo** [comment]
@maxbrunsfeld and I made a bunch more progress on fixing this today. It's a huge change because we're basically overhauling the entire document model, but I am really optimistic for the new design. Work is taking place in [atom/text-document](https://github.com/atom/text-document). It's going to be at least a few weeks until it's ready though. Hang in there.

**Comment by russlescai** [comment]
Awesome, let us know if there is anything we can test to help out. :)

**Comment by karai17** [comment]
The hero we needed.

By overhauling the DOM, do you think it will ultimately speed up Atom as a whole? One of my only real gripes with Atom is how it is substantially slower than other text editors. If this fix could bring it up to, or near speed parity with others, it would catapult Atom way ahead of the rest in terms of usability and functionality.

**Comment by benogle** [comment]
> overhauling the entire document model

I guess `document model` is ambiguous in this context. This doesnt touch the DOM, just the underpinnings of how the buffer is stored and referenced.

**Comment by karai17** [comment]
Ah, well, I guess the same question applies. Are these changes intended to speed up Atom as a whole?

**Comment by nathansobo** [comment]
That is the plan. I predict it will help us a lot but we may also need to continue optimizing at other layers as well. It's a big piece of the puzzle though.

**Comment by karai17** [comment]
Awesome! I look very forward to this.

**Comment by octref** [comment]
@karai17 
Just wondering. You mentioned you need to edit 3D model files. How does those files look like?
And, as for syntax-highlighting on them, do you
- Absolutely need it
- Don't want it at all
- Can live without it but wish to have it

**Comment by karai17** [comment]
@octref 

[here](https://raw.githubusercontent.com/excessive/love-und-panzer/master/assets/models/tank.iqe) is an example of a several megabyte IQE 3D model file. The weird indenting is caused by Blender's IQE exporter, it is not part of the spec.

Since IQE is a data format and not a language format, I wouldn't really expect it to have syntax highlighting. Maybe it would be neat to write a package that did highlight it, but I don't think it would be all that useful.

There are some language files that can reach huge sizes such as SQL dumps that would benefit from syntax highlighting, though.

**Comment by octref** [comment]
@karai17 

Got it.
Currently Atom is using [converted textmate grammar bundles](https://github.com/atom/language-coffee-script). Once the support for large file is there, you should be able to write a custom syntax package and get syntax-highlighting in Atom for the 3D model files.

**Comment by karai17** [comment]
Awesome! One thing I noticed recently is that, at least in the `language-lua` package, some syntax is highlighted in a way that I don't believe it should be, and I am not sure if that is the package itself or a symptom of how Atom colours syntax.

An example would be that if I use a Lua core function such as `print()`, Atom highlights that text. But if I type it incorrectly, such as `pirnt()`, it still highlights it. What seems to be happening is the package picks up on `()` and highlights it as a function. With sublime I believe only core functions were highlighted, or at the very least core and user functions had separate colours.

I know that this falls outside of the scope of this particular issue, but I was wondering if it had to do with Atom's syntax highlighting or if it was just an oversight on the package itself?

**Comment by LPGhatguy** [comment]
I've seen some interesting behavior with themes too as far as Lua core functions go.

Here's a side-by-side comparison of some code. Sublime Text 3 is on the left (correct behavior), while Atom is on the right, using the same theme as ST3 using the theme importer from a couple versions back.

http://i.imgur.com/V9b5GWh.png

**Comment by octref** [comment]
@karai17 
If you write a `test.lua` file with the following two lines:
```
print()
pirnt()
```
Use `Ctrl-Shift-I` to open the inspector, and `Ctrl-F` find the corresponding DOM node:
For `print()`, it's:
`<span class="support function lua">print</span>`
For `pirnt()`, it's:
`<span class="support function any-method lua">pirnt</span>`
On the inspector, you can find they are all colored by the `.support.function` class, which corresponds to this LESS rule in Atom-Dark theme:
https://github.com/atom/atom-dark-syntax/blob/master/index.less#L218-L220
The grammar:
https://github.com/FireZenk/language-lua/blob/d1bbd1805cfc1e0cfaeb60b426133f72b442d2e3/grammars/lua.cson#L131-L133

So if you are unhappy about this grammar syntax, you can open an issue to `language-lua`, asking them to parse `print` differently.

For example, python parses `print` as `keyword.other.python` and gives it a pretty azure :-)
`<span class="keyword other python">print</span>`

### But
@karai17 , @LPGhatguy
Let's stop the conversation here. Many people are subscribed to this issue to get progress on large file support.

**Comment by erikdonohoo** [comment]
Super happy to hear this is being worked on.  Its not just large binary files that are issues here.  Any js framework (like angular, ember) end up being too large for atom to handle.  It doesn't seem like an edge case to occasionally need to look at the source of some 3rd party package you are using.

**Comment by wpostma** [comment]
This editor is a toy, if it can't handle large files.    I'll be watching to see if Atom matures in this area.

**Comment by murphyj** [comment]
Agree with @erikdonohoo - this is not an edge case. You also can't open some crucial files for Ghost because it's ember.js backed (hence has large binary vendor files).

**Comment by sduensin** [comment]
This is a HUGE problem.  2 meg is nothing these days.  As much as I like Atom, this limitation makes it pretty useless for day to day work.  :-(

**Comment by sceee** [comment]
I can support this - many files (let it be log files etc.) have more than 2 MB in size. This always makes you need another editor to open these files beside atom.

**Comment by gregzx** [comment]
This limitation is really a dealbreaker when it comes to using atom... I hope this gets improved one day.

**Comment by karlbateman** [comment]
This should be considered a standard feature for a modern text editor. I would really love to see this feature added as Atom is such an amazing text editor.

**Comment by 50Wliu** [comment]
This is currently being worked on.  The Atom team recognizes that it's a big problem, but it also needs a big rewrite in order to make everything work.  See https://github.com/atom/text-document for more details.

**Comment by karlbateman** [comment]
@50Wliu I can appreciate the work involved. Looking forward to seeing this in a future release :smile:

**Comment by backspaces** [comment]
Could we pin this? The extraordinary atom/text-document should be shown to
all those concerned.

I say this because Atom is historic, like several JS/Html/Browser/Node
projects have been over the last few years.

Many of Atom's clients are simply looking for a coherent text editor.
That's fine.  But they should also understand the JavaScript Revolution
background.

**Comment by Tapefabrik** [comment]
+1

**Comment by mika76** [comment]
I must admit these days the only thing keeping me from using Atom is the performance and the large file non-support. I can't wait to see this working :grin:

**Comment by xpepermint** [comment]
+1

**Comment by theksmith** [comment]
+100

**Comment by pixelchutes** [comment]
+ 2.0001 (MB) ...wait for it... :wink:

**Comment by thedaniel** [comment]
This is actively being worked on and is a high priority, no reason to post +1 comments anymore - it just spams everyone watching the issue.

**Comment by matmuchrapna** [comment]
@thedaniel issues can be closed for commenting

**Comment by ilanbiala** [comment]
@thedaniel I think it would help if any other possible progress updates were provided. If there aren't any, it would just be nice to know that it's happening by @nathansobo or you commenting in addition to adding the in progress label.

**Comment by nathansobo** [comment]
It's happening. Check out the text-document repo. Out of the country for a conference which is consuming some attention, but it's been almost my sole focus. Not much more to say than the progress and write-ups you can observe there. I empathize with everyone who wants this because I want it too, and it's been challenging but we're definitely getting there.

**Comment by robhawkes** [comment]
@thedaniel @matmuchrapna: Yup, it's [simple to lock the comments](https://help.github.com/articles/locking-conversations/) while still allowing contributors to discuss things.

**Comment by kylegoetz** [comment]
I disagree this is an edge case. There are popular third-party JS libraries that are greater than this limit. What motivates my finding this bug is that Telerik's Kendo library is too big to open in Atom unless you open the minified version, which is, of course, not useful at all.

I'm having to read the library in Firefox's debugger since I can't open it in Atom to read it. You can argue philosophy about whether I should be opening a file I don't intend to edit in Atom, but I might edit it if I need to and just consider it no longer a proper third party lib but rather a first party, but can't edit in Atom.

**Comment by benogle** [comment]
I've locked this. We do believe this is an important case to handle. We are currently working on this. Please visit https://github.com/atom/text-document for more information.

**Comment by nathansobo** [comment]
Just an update here... in the interest of time we've decided to refocus our efforts on more incremental improvements.

* One large chunk of work from `text-document`, the `MarkerIndex`, is in the process of being integrated into our existing document model. This will improve Atom's performance in the presence of large numbers of markers, such as when visualizing numerous search results in large files. The [pull request](https://github.com/atom/text-buffer/pull/58) should be merged in the next few days.

* We're also working to [reduce the memory consumption](https://github.com/atom/atom/pull/6757) of our tokenized line representation.

We'll be searching for several more smaller wins like this, then redirect our attention back to some of the bigger changes researched in `text-document`. Making Atom performant for files of all sizes remains a top priority. We'll continue to post updates here.

**Comment by nathansobo** [comment]
There's still more to do here, but a basic fix for the 80% case of loading large files is now on master. Before 1.0 I'd like to fix a few performance hiccups when moving the cursor in huge files. Post 1.0 I'd like to:

* Syntax highlight in a background process so we can highlight huge files without hiccups.
* Show a progress bar while performing I/O and computing initial metadata while loading a large file.
* Drop the memory overhead another order of magnitude.
* Support folds in large files

We can create separate issues for all of those, but I'm going to close this one. Atom can now load and edit large files. Don't worry, we'll keep working on refining it.
